### ISSUE TITLE
Human friendly name generator for default study names.

### ISSUE THREAD

**Issue by toshihikoyanase** [optuna/optuna#803]
## Motivation

This issue was originally mentioned at https://github.com/optuna/optuna/pull/799#discussion_r359110361 by @sile.

The format of the current default study name is `no-name-` + `UUID`. It is sometimes difficult for users to distinguish each study. For example, we can only see the prefix of the study name like `no-name-0` in the following screenshot.

![image](https://user-images.githubusercontent.com/3255979/71048071-5f349200-2181-11ea-9ff9-eb2d3c8af060.png)

I think we can use more human-friendly name generators like Docker.

## Description

Current name generator code for `RDBStorage` is [here](https://github.com/optuna/optuna/blob/2c89095e931ad9e2ed4df887cd6e50002a0a6c33/optuna/storages/rdb/storage.py#L178-L188).

In my quick survey, Docker seems to employ [moby](https://github.com/moby/moby) for the name generation and it manages [some dictionaries](https://github.com/moby/moby/blob/master/pkg/namesgenerator/names-generator.go). moby is ported to python as [moby-dict](https://pypi.org/project/moby-dict/), and we may be able to use it. But I'm not sure if moby-dict is the best solution because it is not actively maintained as far as I see [its release history](https://pypi.org/project/moby-dict/#history). So, I think we need a further survey and design discussion.

**Comment by ghost2718** [comment]
Hi, Can I work on this issue

**Comment by toshihikoyanase** [comment]
@ghost2718 Thank you for your interest.

If you work on it, [`mong`](https://github.com/toshihikoyanase/mong), a docker-like name generator, may be useful.
I'm developing it in my personal repository, and one of the features is the up-to-date name dictionary using the [GitHub Actions workflow](https://github.com/toshihikoyanase/mong/actions?query=workflow%3A%22check+dict+update%22).  We can see some existing libraries (e.g., https://github.com/shamrin/namesgenerator), but they are not developed actively.

So, I hope it can ease the implementation of this feature. If you find the mong's feature is not enough for this issue, it's OK to replace it with other implementation.

**Comment by ghost2718** [comment]
@toshihikoyanase Thank you, I will start working on this issue. I will actively seek your advice regarding any implementation probem.

**Comment by toshihikoyanase** [comment]
@ghost2718 Thank you! If you have any questions, please let me know.

**Comment by ghost2718** [comment]
@toshihikoyanase I have looked at other name generators, along with the one you had suggested. I felt that more or less every one of them is like mong. So I tried this out:

`    def _create_unique_study_name(session):
        # type: (orm.Session) -> str

        while True:
            # study_uuid = str(uuid.uuid4())
            # study_name = DEFAULT_STUDY_NAME_PREFIX + study_uuid
            study_name = mong.get_random_name()
            study = models.StudyModel.find_by_name(study_name, session)
            if study is None:
                break

        return study_name
`

Is it okay? or am I missing something?

**Comment by toshihikoyanase** [comment]
@ghost2718 Thank you for your proposal.

I think we need to add suffixes to the name generated by mong. This is because the number of unique names is about 25,000 and it seems to be too small for some intensive users. What do you think if you add 3 or 4 random digits after the name generated by mong?

In addition, `while True` may cause an infinite loop if candidate names are not available. How about setting the maximum retry count?

**Comment by hvy** [comment]
We could also add prefixes instead which might allow some sort of sorting. Instead of random number we could then use formatted dates or Unix times. It might also be reasonable not being too generous in the name generator and recommend intensive users to actually specify names explicitly.

**Comment by sile** [comment]
>  Instead of random numbers we could then use formatted dates or Unix times.

Agree. I personally prefer human understandable datetime format to unixtime format.

>  It might also be reasonable not being too generous in the name generator and recommend intensive users to actually specify names explicitly.

This choice also sounds acceptable. I think that we could raise an exception and recommend the user to specify the name manually if the retry number of a name generation exceeded a predefined count.

**Comment by ghost2718** [comment]
Hello @toshihikoyanase , sorry for the delay. I read the above suggestions. I personally felt that a four-digit code as a prefix would make them look cleaner. Anyways I edited the code like this.
 `     
  def _create_unique_study_name(session):
       
        create_time = datetime.now()
        prefix = create_time.strftime("%m/%d/%Y-%H:%M:%S")
        num_tries = 0
        while True and num_tries < 1000:
            # study_uuid = str(uuid.uuid4())
            # study_name = DEFAULT_STUDY_NAME_PREFIX + study_uuid
            study_name = mong.get_random_name()
            study_name = prefix + study_name
            study = models.StudyModel.find_by_name(study_name, session)
            if study is None:
                break
            num_tries += 1
        if num_tries < 1000:
            return study_name
        else:
            print("Please Manually set the Name of the study.") `

Please point out the suggestions or improvements I have to make.

Thanks

**Comment by toshihikoyanase** [comment]
Hello @ghost2718 , thank you for including the suggestions.

> I personally felt that a four-digit code as a prefix would make them look cleaner.

Hmm, everyone has their own preference in the style of `prefix`. So, how about removing the prefix and creating a PR which contains only minimum code?

Through the discussion, I found that the main discussion points are as follows:

1. Format of the default names
2. Exception handling of name collision

## 1. Format of the default names

I think the docker-like name style is the simplest among the candidates in this issue. If we just use the style, we can discuss the prefix in a follow-up PR.

## 2. Exception handling of name collision

As @sile said, we need to take care of name collision. 

> I think that we could raise an exception and recommend the user to specify the name manually if the retry number of a name generation exceeded a predefined count.

If we employ the retry mechanism in your example code, we still handle the case where `num_tries >= 1000`. So, such error handling seems to be mandatory.

## An example code:

In summary, I think code will be as follows:

```python
def _create_unique_study_name(session):

    # study_uuid = str(uuid.uuid4())
    # study_name = DEFAULT_STUDY_NAME_PREFIX + study_uuid

    # 1. Format of the default names
    study_name = mong.get_random_name()
    study = models.StudyModel.find_by_name(study_name, session)

    # 2. Exception handling of name collision
    if study is not None:
        # TODO: Please revise the error type and the error message.
        raise ValueError("Failed to generate study name automatically. Please set the study name manually.")

```

For further discussion, could you create a PR? I think we can discuss easier using the interface of GitHub's Pull Request.

**Comment by toshihikoyanase** [comment]
This is still an open issue.

@va1shn9v Please let me know if you still work on it.

**Comment by hvy** [comment]
Closed since the priority is considered low now that it's been stale for a while. Feel free to reopen as needed.
